providers/dns/gravity/gravity.go

```diff
@@ -0,0 +1,207 @@
+func NewDNSProvider() (*DNSProvider, error) {
+	values, err := env.Get(EnvUsername, EnvPassword, EnvServerURL)
+	if err != nil {
+		return nil, fmt.Errorf("gravity: %w", err)
+	}
+
+	config := NewDefaultConfig()
+	config.Username = values[EnvUsername]
+	config.Password = values[EnvPassword]
+	config.ServerURL = values[EnvServerURL]
+
+	return NewDNSProviderConfig(config)
+}
+
+// NewDNSProviderConfig return a DNSProvider instance configured for Gravity.
+func NewDNSProviderConfig(config *Config) (*DNSProvider, error) {
+	if config == nil {
+		return nil, errors.New("gravity: the configuration of the DNS provider is nil")
+	}
+
+	client, err := internal.NewClient(config.ServerURL, config.Username, config.Password)
+	if err != nil {
+		return nil, fmt.Errorf("gravity: %w", err)
+	}
+
+	if config.HTTPClient != nil {
+		client.HTTPClient = config.HTTPClient
+	}
+
+	client.HTTPClient = clientdebug.Wrap(client.HTTPClient)
+
+	return &DNSProvider{
+		config:  config,
+		client:  client,
+		records: make(map[string]internal.Record),
+	}, nil
+}
+
+// Present creates a TXT record using the specified parameters.
+func (d *DNSProvider) Present(domain, token, keyAuth string) error {
+	ctx := context.Background()
+
+	info := dns01.GetChallengeInfo(domain, keyAuth)
+
+	_, err := d.client.Login(ctx)
+	if err != nil {
+		return fmt.Errorf("gravity: login: %w", err)
+	}
+
+	zone, err := d.findZone(ctx, info.EffectiveFQDN)
+	if err != nil {
+		return fmt.Errorf("gravity: %w", err)
+	}
+
+	subDomain, err := dns01.ExtractSubDomain(info.EffectiveFQDN, zone)
+	if err != nil {
+		return fmt.Errorf("gravity: %w", err)
+	}
+
+	id, err := uuid.NewV4()
+	if err != nil {
+		return err
+	}
+
+	record := internal.Record{
+		Data:     info.Value,
+		Hostname: subDomain,
+		Type:     "TXT",
+		UID:      id.String(),
+	}
+
+	err = d.client.CreateDNSRecord(ctx, zone, record)
+	if err != nil {
+		return fmt.Errorf("gravity: create DNS record: %w", err)
+	}
+
+	d.recordsMu.Lock()
+
+	record.Fqdn = zone
+	d.records[token] = record
+	d.recordsMu.Unlock()
+
+	return nil
+}
+
+// CleanUp removes the TXT record matching the specified parameters.
+func (d *DNSProvider) CleanUp(domain, token, keyAuth string) error {
+	info := dns01.GetChallengeInfo(domain, keyAuth)
+
+	d.recordsMu.Lock()
+	record, ok := d.records[token]
+	d.recordsMu.Unlock()
+
+	if !ok {
+		return fmt.Errorf("gravity: unknown record for '%s' '%s'", info.EffectiveFQDN, token)
+	}
+
+	err := d.client.DeleteDNSRecord(context.Background(), record.Fqdn, record)
+	if err != nil {
+		return fmt.Errorf("gravity: delete record: %w", err)
+	}
+
+	return nil
+}
+
+// Timeout returns the timeout and interval to use when checking for DNS propagation.
+// Adjusting here to cope with spikes in propagation times.
+func (d *DNSProvider) Timeout() (timeout, interval time.Duration) {
+	return d.config.PropagationTimeout, d.config.PollingInterval
+}
+
+// Sequential All DNS challenges for this provider will be resolved sequentially.
```

Comment:
Go doc comments sometimes require contorted sentences. How about this wording?

```suggestion
// Sequential implements the [dns01.sqeuential] interface.
// It changes the behavior of the provider to resolve all DNS challenges sequentially.
```